Add kernel sync github action

bmastbergen opened the pull request:
This PR adds a github action to sync the kernel-mainline, stable_6.12.y, and centos9 branches with their respective upstreams.  The sync happens daily at 8am (an arbitrary choice of time) or manually.  This should be easily extensible by adding new local branch, remote, and remote branch information to the build matrix (see the 2nd and 3rd commit that add stable_6.12.y and centos9 as examples of that).

Initially I thought that each branch would have its own github action, but then I realized that adding the github action would mess up the history and would mean the branches were not identical to their upstream.  So I ended up adding a github action to the main branch (similar to how [kernels-upstream](https://github.com/ctrliq/kernels-upstream/blob/main/.github/workflows/kernel.yml) does things).

I didn't want to let this loose on our actual branches so I've done a bit of testing on a fork of the repo to get warm fuzzies that it does the right thing.  That being said I will monitor things for a while after this is merged to make sure everything is happening correctly.

LE-2708
LE-2709
LE-2710

PlaidCat commented:
> I didn't want to let this loose on our actual branches so I've done a bit of testing on a fork of the repo to get warm fuzzies that it does the right thing. 

Do you have any evidence?


I've added a couple of comments, and I was expecting to see these on the branches themselves but this might be fine as well since its not related to the reset of the branches.

bmastbergen commented:
> > I didn't want to let this loose on our actual branches so I've done a bit of testing on a fork of the repo to get warm fuzzies that it does the right thing.
> 
> Do you have any evidence?
> 
> I've added a couple of comments, and I was expecting to see these on the branches themselves but this might be fine as well since its not related to the reset of the branches.

Here is a test run on my fork:
https://github.com/bmastbergen/kernel-src-tree/actions/runs/14333423381

And the resultant branches:
https://github.com/bmastbergen/kernel-src-tree/branches

I don't see any comments, so maybe you need to publish your review?  Thanks!

bmastbergen commented on .github/workflows/sync.yml:
```diff
@@ -0,0 +1,57 @@
+
+jobs:
+  build:
+    runs-on: kernel-build-lol
```
Ah yes.  I copied that from kernels-upstream, but that action actually does a kernel build.  Will change.

bmastbergen commented on .github/workflows/sync.yml:
```diff
@@ -0,0 +1,57 @@
+      image: rockylinux:9
+    strategy:
+      fail-fast: false
+      matrix:
```
Pretty much.  A job is spawned per matrix element.  If you look at the example action below you'll see that three jobs run, each with the unique branch/remote/remote branch environment vars.

https://github.com/bmastbergen/kernel-src-tree/actions/runs/14333423381

bmastbergen commented on .github/workflows/sync.yml:
```diff
@@ -0,0 +1,57 @@
+      REMOTE_BRANCH: ${{ matrix.remote_branch }}
+
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v4
```
Hmm, yea that might actually work.  Again, I did what kernels-upstream does today, but this might be a better way.  I'll mess with it a bit more.  Thanks!

wizctrl commented on .github/workflows/sync.yml:
```diff
@@ -0,0 +1,57 @@
+      matrix:
+         include:
+           - branch: kernel-mainline
+             remote: git://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git
```
```suggestion
             remote: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git
```

wizctrl replied:
Adding https instead of git protocol

wizctrl commented on .github/workflows/sync.yml:
```diff
@@ -0,0 +1,57 @@
+             remote: git://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git
+             remote_branch: master
+           - branch: stable_6.12.y
+             remote: git://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git
```
```suggestion
             remote: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git
```

wizctrl replied:
Adding https instead of git protocol

wizctrl commented on .github/workflows/sync.yml:
```diff
@@ -0,0 +1,57 @@
+
+      - name: Install git
+        run: |
+          dnf install -y git
```
If `runs-on` above is changed to `ubuntu-latest` then we will have to update dnf stuff

PlaidCat replied:
its actually running in a Rcoky Container so this works, we use it throughout this repo

wizctrl replied:
Ah I missed the container field flag, makes sense 👍

wizctrl commented on .github/workflows/sync.yml:
```diff
@@ -0,0 +1,57 @@
+
+          git remote add upstream ${REMOTE}
+          git fetch upstream --tags --quiet
+          git rebase upstream/${REMOTE_BRANCH}
```
```suggestion
          if ! git rebase "upstream/${REMOTE_BRANCH}"; then
            echo "Error: Rebase failed on branch 'upstream/${REMOTE_BRANCH}', likely due to conflicts"
            git rebase --abort
            exit 1
          fi
```

Added some error handling, rebasing makes me nervous sometimes

bmastbergen replied:
We have 'set -e' so the script is already going to stop immediately if 'git rebase' fails, right?

PlaidCat replied:
I usually prefer doing the 
```
RET=$()
if RET -ne 0 
```

But this is a VERY C thing todo

PlaidCat replied:
I don't tend to use `set -e` so I didn't even think about it.

This is fine by me

wizctrl reviewed (commented):
Hi Brett, hate to be a stickler, but added some comments. Please let me know what you think about them

PlaidCat commented on .github/workflows/sync.yml:
```diff
@@ -0,0 +1,57 @@
+          git remote add upstream ${REMOTE}
+          git fetch upstream --tags --quiet
+          git rebase upstream/${REMOTE_BRANCH}
+          git push origin --follow-tags
```
should we explicitly add the `${BRANCH}` name here ?

wizctrl replied:
How exactly do you mean?

PlaidCat replied:
`git push origin --follow-tags` -> `git push origin --follow-tags ${BRANCH}`

bmastbergen commented:
@PlaidCat @wizctrl

Just force pushed.  I think I've addressed most of your comments:
- switched to ubuntu-latest
- switch to https for remotes
- actually using action/checkout@v4 to check out the target branch

See what you think.  Thanks!

wizctrl commented:
If you and Maple prefer keeping it simple by skipping the additional rebase checks for now, this looks good to me! Thanks, Brett.
